misc/python/materialize/checks/multiple_partitions.py

```diff
@@ -0,0 +1,102 @@
+           "type": "record",
+           "name": "Key",
+           "fields": [
+               {"name": "key1", "type": "string"}
+           ]
+         }
+
+       $ set schema={
+           "type" : "record",
+           "name" : "test",
+           "fields" : [
+               {"name":"f1", "type":"string"}
+           ]
+         }
+    """
+    )
+
+class MultiplePartitions(Check):
+    """Test that adds new partitions to a Kafka source"""
+
+    def initialize(self) -> Testdrive:
+        return Testdrive(
+            schemas()
+            + dedent(
+                """
+                $ kafka-create-topic topic=movies-topic
+
+                $ kafka-ingest format=avro key-format=avro topic=movies-topic key-schema=${keyschema} schema=${schema} repeat=1000
+                {"key1": "A${kafka-ingest.iteration}"} {"f1": "A${kafka-ingest.iteration}"}
+
+                > CREATE CONNECTION IF NOT EXISTS kafka_conn FOR KAFKA BROKER '${testdrive.kafka-addr}';
+
+                > CREATE CONNECTION IF NOT EXISTS csr_conn FOR CONFLUENT SCHEMA REGISTRY URL '${testdrive.schema-registry-url}';
+
+                > CREATE SOURCE movies_source
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-movies-topic-${testdrive.seed}', TOPIC METADATA REFRESH INTERVAL MS 500)
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE UPSERT
+
+                > CREATE MATERIALIZED VIEW mv_movies AS SELECT * FROM movies_source;
+                
+                $ kafka-add-partitions topic=movies-topic total-partitions=4
+                """
+            )
+        )
+
+    def manipulate(self) -> List[Testdrive]:
+        return [
+            Testdrive(schemas() + dedent(s))
+            for s in [
+                """
+                $ kafka-ingest format=avro key-format=avro topic=movies-topic key-schema=${keyschema} schema=${schema} repeat=40
+                {"key1": "A${kafka-ingest.iteration}"} {"f1": "A${kafka-ingest.iteration}"}
+                """,
+                """
+                $ kafka-ingest format=avro key-format=avro topic=movies-topic key-schema=${keyschema} schema=${schema} repeat=60
+                {"key1": "A${kafka-ingest.iteration}"} {"f1": "A${kafka-ingest.iteration}"}
+                """,
+            ]
+        ]
+
+    def validate(self) -> Testdrive:
+        return Testdrive(
+            dedent(
+                """
+                > SELECT * FROM movies_source_progress;
+                (3,) 0
+                [0,0] 1025
+                [1,1] 25
+                [2,2] 25
+                [3,3] 25
+                
+                > SELECT status FROM mz_internal.mz_source_statuses WHERE name = 'movies_source';
+                running
+                
+                # > SELECT COUNT(*) FROM movies_source;
+                # 1100
+                
+                # > SELECT COUNT(*) FROM mv_movies;
+                # 1100
```

Comment:
I have a couple of questions:

1. The sources and materialized views still return `COUNT(*)` = 1000 but I would expect 1100. Why is this? This does not change if I 
    - wait for more than five minutes using `sleep-is-probably-flaky-i-have-justified-my-need`
    - create a new source
    - create everything new including the connection
2. Unlike `mz_internal.mz_source_statuses`, `mz_internal.mz_source_statistics` does not contain any entries. Is this expected and do the statistics need to be actived?
3. `movies_source_progress` sees 1100 entries. Does this mean that `TOPIC METADATA REFRESH INTERVAL MS 500)` is effective such that the comment `# gh#12193 : topic_metadata_refresh_interval_ms is not observed so a default refresh interval of 300s applies` in `test/limits/mzcompose.py` is outdated and issue https://github.com/MaterializeInc/materialize/issues/12193 is obsolete? Or is this related to why 1. produces surprising results (though, I was waiting also for more than five minutes)?
4. Is there anything else that should be validated?

Reply:
> The sources and materialized views still return COUNT(*) = 1000 but I would expect 1100. Why is this? This does not change if I

This is an upsert source, so newer entries using the same key overwrite previous entries. All your entries have keys starting with `A`, so they will overwrite each other. Ideally, to stress the entire thing to the max, you would want to have both old keys (that is , upsert updates) and entirely new keys (that is, upsert inserts) at every ingestion step.

Also, when you design your validation queries, make sure to design them so that they will fail if the updates are not properly processed. A simple `COUNT(*)= 1000` over an upsert source will still succeed even if all the updates have been thrown away. `alter_index.py` for example uses `GROUP BY` in the `validate()` . Another approach is to use `MAX()` in your query so that you know that the source has been updated to the latest value expected.

> Unlike mz_internal.mz_source_statuses, mz_internal.mz_source_statistics does not contain any entries. Is this expected and do the statistics need to be actived?

This should work, at least according to `test/testdrive/source-statistics.td` . That file however has a comment that it takes up to 30 seconds for the statistics to update, which would be I think too long of a wait for Platform Checks, so if this is indeed the case, do not query that table in this test.

> movies_source_progress sees 1100 entries. Does this mean that TOPIC METADATA REFRESH INTERVAL MS 500) is effective such that the comment # gh#12193 : topic_metadata_refresh_interval_ms is not observed so a default refresh interval of 300s applies in test/limits/mzcompose.py is outdated and issue https://github.com/MaterializeInc/materialize/issues/12193 is obsolete?

This depends on how you ran the test . Did you use `--scenario=RestartEntireMz`? If yes, then I think the metadata is being refreshed once every restart. But if you ran it with `--scenario=NoRestartNoUpgrade`, Mz is not restarted, so you will see the default metadata refresh interval of 300s kick in . I think.

> Is there anything else that should be validated?

You also need to ingest some upsert deletions as you go -- those are `kafka-ingest` entries that only have a key but no payload.

Reply:
Thanks a lot for your answers!

As for the metadata refresh, I was using `scenario=NoRestartNoUpgrade`. I will keep an eye on it.

Reply:
You were absolutely right about `source-statistics.td`. It takes some time until this table is populated.